woocommerce/woocommerce-square
# Fix potential infinite loop issue with pull_inventory step in Manual synchronization

### All Submissions:



---

### Changes proposed in this Pull Request:
The PR fixes a potential issue with the pull_inventory step, which could cause the step to run repeatedly until we manually stop or complete the sync. This is an edge case and is not reproducible using UI steps. However, I noticed a condition in the  [code](https://github.com/woocommerce/woocommerce-square/blob/trunk/includes/Sync/Manual_Synchronization.php#L1508) where we check if the product exists for the Square variation before updating inventory. In the edge case where the product is permanently deleted while the Square sync is running, and the deleted product is already recorded as a validated product in the sync, the pull_inventory step doesn't mark the product as processed. As a result, the step continues to run repeatedly until we manually stop or complete the sync.

This PR adds handling for this case and includes an alert for the deleted product, so the store owner is informed about the error and the sync step doesn't get stuck.

![image](https://github.com/user-attachments/assets/10b638d6-1623-432f-94d7-de6101db2e4f)


Closes #84 

### Steps to test the changes in this Pull Request:
This is an edge case, and I am not able to reproduce the issue using UI steps. However, you can simulate it with the following steps:
1. Configure the Square settings with Inventory sync enabled. (Keep SOR as you prefer)
1. Sync the product once.
1. Add the code snippet below to the theme's `functions.php` file (this is for simulating the edge case) and replace PRODUCT_ID with one of the synced product ID.
```php
add_action( 'wc_square_background_sync_job_updated', function ( $job ) {
	if ( $job && $job->next_steps && 'pull_inventory' === $job->next_steps[0] ) {
		wp_delete_post( PRODUCT_ID, true );
	}
} );

```
4. Start manual synchronization.
5. Verify that Sync didn't get stuck and completed successfully. Also, you will see a similar alert given in the screenshot above.
### Changelog entry
> Fix - Address a potential infinite loop issue with the `pull_inventory` step when running a manual synchronization.
